java/src/main/native/ai_onnxruntime_OrtTrainingSession.c

```diff
@@ -0,0 +1,745 @@
+/*
+ * Copyright (c) 2022 Oracle and/or its affiliates. All rights reserved.
+ * Licensed under the MIT License.
+ */
+#include <jni.h>
+#include <string.h>
+#include <stdlib.h>
+#include "OrtJniUtil.h"
+#include "onnxruntime/core/session/onnxruntime_c_api.h"
+#include "onnxruntime_training_c_api.h"
+#include "ai_onnxruntime_OrtTrainingSession.h"
+
+/*
+ * Class:     ai_onnxruntime_OrtTrainingSession
+ * Method:    createTrainingSession
+ * Signature: (JJJJJLjava/lang/String;Ljava/lang/String;Ljava/lang/String;)J
+ */
+JNIEXPORT jlong JNICALL Java_ai_onnxruntime_OrtTrainingSession_createTrainingSession
+  (JNIEnv * jniEnv, jclass clazz, jlong apiHandle, jlong trainApiHandle,
+     jlong envHandle, jlong optionsHandle, jlong checkpointHandle,
+     jstring trainPath, jstring evalPath, jstring  optimizerPath) {
+  (void) clazz; // Required JNI parameters not needed by functions which don't need to access their host class.
+  const OrtApi* api = (const OrtApi*) apiHandle;
+  const OrtTrainingApi* trainApi = (const OrtTrainingApi*) trainApiHandle;
+  const OrtEnv* env = (const OrtEnv*) envHandle;
+  const OrtSessionOptions* options = (const OrtSessionOptions*) optionsHandle;
+  OrtCheckpointState* checkpoint = (OrtCheckpointState*) checkpointHandle;
+
+  jsize evalLength = (*jniEnv)->GetStringLength(jniEnv, evalPath);
+  jsize optimizerLength = (*jniEnv)->GetStringLength(jniEnv, optimizerPath);
+  OrtTrainingSession* session = NULL;
+
+#ifdef _WIN32
+  // The output of GetStringChars is not null-terminated, so we copy it and add a terminator
+  const jchar* trainJavaStr = (*jniEnv)->GetStringChars(jniEnv, trainPath, NULL);
```

Comment:
>GetStringChars [](http://example.com/codeflow?start=41&length=14)

Can GetStringChars() fail?

Reply:
Not unless the pointer is null, which I'll add a check for in Java.

Reply:
There already is a null check in Java for `trainPath` and I refactored the logic for the other two.